lib/puppet/provider/service/systemd.rb

```diff
@@ -28,12 +28,33 @@ def disable
 
   def enabled?
     begin
-      systemctl("is-enabled", @resource[:name])
+      systemctl_info = systemctl(
+         'show',
+         @resource[:name],
+         '--property', 'LoadState',
+         '--property', 'UnitFileState',
+         '--no-pager'
+      )
+
+      svc_info = Hash.new
+      systemctl_info.split.map{|svc|
+        entry_pair = svc.split('=')
+        svc_info[entry_pair.first.to_sym] = entry_pair.last
+      }
+
+      # The masked state is equivalent to the disabled state in terms of
+      # comparison so we only care to check if it is masked if we want to keep
+      # it masked.
+      #
+      # We only return :mask if we're trying to mask the service. This prevents
+      # flapping when simply trying to disable a masked service.
+      return :mask if (@resource[:enable] == :mask) && (svc_info[:LoadState] == 'masked')
```

Comment:
In the case of Puppet enabled? returns a symbol and, in this particular case, it is *already* not a boolean.

Valid values are :true, :false, :manual, :mask (new)

If :true then true, else false. In this case :mask is equivalent to :false since it is not :true.

Speaking of unexpected :true != true and :false != false has caused no end of confusion with writing custom types.

Honestly, to do anything else would make the maintainability of the code MUCH more difficult.